### PR TITLE
Fix #5787: Send to ENS wallet address

### LINKED ISSUE #5787: Wallet: Support sending transactions to a domain (ENS)
### Description: 

Update send transaction view / logic to support sending to a domain.

ref / more info: 
https://bravesoftware.slack.com/archives/C023VS4HJ6Q/p1659646211852199
https://github.com/brave/brave-browser/issues/18681
https://github.com/brave/brave-browser/issues/18656

### PR DESCRIPTION
**Pull request by StephenHeaps**
## Summary of Changes
- Support ENS wallet address resolution in Send token
- Includes ENS offchain wallet address resolution, which defaults to asking the user before submitting request offchain.
- Error copy was updated for SNS and ENS to be dynamic: https://github.com/brave/brave-browser/issues/28041
    - ex. `Domain doesn't have a linked Ethereum address`
- Security / privacy review: https://github.com/brave/security/issues/1206 

This pull request fixes brave/brave-ios#5787

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:
1. Open wallet/web3 settings and verify default setting of `Allow ENS Offchain Lookup` is set to `Ask`
2. Navigate to send token view
3. Select an Ethereum network
4. Enter `ethereum.eth` as a send to address and verify it resolves to the address as expected (with loading state).
5. Enter `offchainexample.eth` as a send address and verify you are prompted to enable ENS offchain lookup.
6. Tap `Use ENS Domain` and verify it resolves to the address as expected (with loading state).
7. Navigate to wallet/web3 settings and verify `Allow ENS Offchain Lookup` is shown as `Enabled`
8. Navigate back to send token view
9. Enter `offchainexample.eth` as a send address and verify you are no longer prompted to enable ENS offchain lookup.
10. Navigate to wallet/web3 settings and verify `Allow ENS Offchain Lookup` is shown as `Disabled`
11. Navigate back to send token view
12.  Enter `offchainexample.eth` as a send address and verify it does not resolve the address

## Screenshots:

https://user-images.githubusercontent.com/5314553/221919341-69fdb98d-f176-4732-9afb-eb78ed1df74e.mp4

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Sources/BraveWallet/WalletStrings.swift [hunk 1]
```diff
@@ -3690,5 +3711,26 @@ extension Strings {
...
+    )
+    public static let ensOffchainGatewayDesc = NSLocalizedString(
+      "wallet.ensOffchainGatewayDesc",
+      tableName: "BraveWallet",
+      bundle: .module,
+      value: "It looks like you've entered an ENS address. We'll need to use a third-party resolver to resolve this request. This helps ensure your .eth domain isn't leaked, and that your transaction is secure.",
+      comment: "Description shown send address / ENS domain when requesting to do an ENS off chain lookup."
+    )
+    public static let ensOffchainGatewayButton = NSLocalizedString(
+      "wallet.ensOffchainGatewayDesc",
```
Comment (nuo-xu) [inline Sources/BraveWallet/WalletStrings.swift]:
```suggestion
      "wallet. ensOffchainGatewayButton",
```

**Review by nuo-xu — APPROVED** [review]

File: Sources/BraveWallet/WalletStrings.swift (hunk 1 above)
Comment (StephenHeaps) [reply Sources/BraveWallet/WalletStrings.swift]: Thanks! Fixed in [fcc221c](https://github.com/brave/brave-ios/pull/7030/commits/fcc221c54b965c6a04da0ca1a0c90a070d1382a2)

File: Sources/BraveWallet/WalletConstants.swift [hunk 2]
```diff
@@ -25,6 +25,9 @@ struct WalletConstants {
   
   /// The url to Brave Help Center for Wallet.
   static let braveWalletSupportURL = URL(string: "https://support.brave.com/hc/en-us/categories/360001059151-Brave-Wallet")!
+  
+  /// The url to learn more about ENS off-chain lookups
+  static let braveWalletENSOffchainURL = URL(string: "https://github.com/brave/brave-browser/wiki/ENS-offchain-lookup")!
```
Comment (stoletheminerals) [inline Sources/BraveWallet/WalletConstants.swift:30]: We should probably add how to enable/disable ENS offchain look up on iOS. Right now it says open brave://settings/web3 which is not available on iOS

Comment (ShivanKaul) [reply Sources/BraveWallet/WalletConstants.swift:30]: Tried to make the page a bit clearer: https://github.com/brave/brave-browser/wiki/ENS-offchain-lookup

**Comment by stoletheminerals** [comment]
It's better to show a relevant error message when offchain look-up is disabled. wdyt @StephenHeaps ?
image

**Comment by StephenHeaps** [comment]
> It's better to show a relevant error message when offchain look-up is disabled. wdyt @StephenHeaps ? 

@stoletheminerals I think this is a good idea. I've opened a core ticket to support this https://github.com/brave/brave-browser/issues/28916 (we need to know if the entered domain requires offchain lookup to show the message so we don't disable all ENS lookups when ENS offchain lookup is disabled), and an iOS ticket to update iOS if this PR has merged prior to the core api update https://github.com/brave/brave-browser/issues/36696

**Review by stoletheminerals — APPROVED** [review]

**Comment by StephenHeaps** [comment]
One minor UI tweak with [iOS 14 support removed](https://github.com/brave/brave-ios/issues/6695) is updating ENS Offchain hyperlink style in Web3 preferences:

Previous | Updated
--|--
221922314-29a670e8-1403-424c-b8c4-aad85cf18611 | Simulator Screen Shot - iPhone 12 Pro - 2023-03-07 at 12 12 27

**Comment by stoletheminerals** [comment]
Maybe `Learn more` instead of `Learn More` since the whole sentence is the link now? 🙂 Would look cleaner imo

**Comment by StephenHeaps** [comment]
> Maybe `Learn more` instead of `Learn More` since the whole sentence is the link now? 🙂 Would look cleaner imo

I agree 🙂 Updated in [4ca083e](https://github.com/brave/brave-ios/pull/7030/commits/4ca083e478e70dba1d726cbbf589af55ae04678b)

**Pull request merged**